sync/src/main/java/tech/pegasys/artemis/sync/SyncManager.java

```diff
@@ -0,0 +1,116 @@
+  private static final int BAD_SYNC_DISCONNECT_REASON_CODE = 128;
+
+  public SyncManager(
+      final Eth2Network network, final ChainStorageClient storageClient, final EventBus eventBus) {
+    this.network = network;
+    this.storageClient = storageClient;
+    this.eventBus = eventBus;
+  }
+
+  private void sync() {
+    Optional<Eth2Peer> possibleSyncPeer = findBestSyncPeer();
+
+    // If there are no peers ahead of us, return
+    if (possibleSyncPeer.isEmpty()) {
+      return;
+    }
+
+    Eth2Peer syncPeer = possibleSyncPeer.get();
+    UnsignedLong syncAdvertisedFinalizedEpoch = syncPeer.getStatus().getFinalizedEpoch();
+
+    CompletableFuture<Void> syncCompletableFuture =
+        requestSyncBlocks(syncPeer, blockResponseListener(syncPeer));
+
+    syncCompletableFuture.thenRun(
+        () -> {
+          if (storageClient
+                  .getStore()
+                  .getFinalizedCheckpoint()
+                  .getEpoch()
+                  .compareTo(syncAdvertisedFinalizedEpoch)
+              < 0) {
+            syncPeer
+                .sendGoodbye(UnsignedLong.valueOf(BAD_SYNC_DISCONNECT_REASON_CODE))
+                .thenRun(this::sync);
+          }
+        });
+  }
+
+  private Optional<Eth2Peer> findBestSyncPeer() {
+    UnsignedLong ourFinalizedEpoch = storageClient.getStore().getFinalizedCheckpoint().getEpoch();
+    return network
+        .streamPeers()
+        .filter(peer -> peer.getStatus().getFinalizedEpoch().compareTo(ourFinalizedEpoch) > 0)
+        .max(
+            (p1, p2) -> {
+              Eth2Peer.StatusData p1StatusData = p1.getStatus();
+              Eth2Peer.StatusData p2StatusData = p2.getStatus();
+              return p1StatusData.getFinalizedEpoch().compareTo(p2StatusData.getFinalizedEpoch());
+            });
+  }
+
+  private CompletableFuture<Void> requestSyncBlocks(
+      Eth2Peer peer, ResponseStream.ResponseListener<BeaconBlock> blockResponseListener) {
+    Eth2Peer.StatusData peerStatusData = peer.getStatus();
+    Bytes32 headBlockRoot = peerStatusData.getHeadRoot();
+    UnsignedLong headBlockSlot = peerStatusData.getHeadSlot();
+    UnsignedLong startSlot =
+        compute_start_slot_at_epoch(storageClient.getStore().getFinalizedCheckpoint().getEpoch());
+    UnsignedLong step = UnsignedLong.ONE;
+    UnsignedLong count = headBlockSlot.minus(startSlot);
+    return peer.requestBlocksByRange(headBlockRoot, startSlot, count, step, blockResponseListener);
+  }
+
+  private ResponseStream.ResponseListener<BeaconBlock> blockResponseListener(Eth2Peer peer) {
+    return ((block) -> {
+      try {
+        Store.Transaction transaction = storageClient.getStore().startTransaction();
+        on_block(transaction, block, stateTransition);
+        transaction.commit();
+        eventBus.post(new StoreDiskUpdateEvent(transaction));
+      } catch (StateTransitionException e) {
+        peer.sendGoodbye(UnsignedLong.valueOf(BAD_SYNC_DISCONNECT_REASON_CODE)).thenRun(this::sync);
```

Comment:
`GoodByeMessage` has some constants we can use for this, you probably want `REASON_FAULT_ERROR`.

Reply:
done.